### PR TITLE
Support mocha-like test files in Imenu

### LINKED ISSUE #575: Imenu support for mocha-like test files
I've created [js2-imenu-mocha](https://github.com/DamienCassou/js2-imenu-mocha) which adds `describe()` and `it()` blocks to imenu entries. Would it make sense to include that in `js2-mode`'s repository directly? Maybe as part of `js2-imenu-extras`?

### PR DESCRIPTION
**Pull request by DamienCassou**
Closes #575.

### CODE REVIEW

File: js2-imenu-extras.el [hunk 1]
```diff
@@ -221,6 +229,161 @@ Currently used for jQuery widgets, Dojo and Enyo declarations."
                                      (list name-value))
                                    (js2-node-abs-pos methods))))))
 
+(defun js2-imenu-record-mocha-describe ()
+  "Populate `js2-imenu-recorder' with mocha-like describe/it/beforeEach/… nodes."
+  (let ((node (js2-node-at-point (1- (point)))))
+    (when (js2-imenu-extras--mocha-top-level-describe-p node)
+      (js2-imenu-extras--mocha-visit-node node (list)))))
```
Comment (dgutov) [inline js2-imenu-extras.el:236]: FWIW, `(list)` evaluates to `nil`.

**Comment by dgutov** [comment]
Solid effort :+1:. Could you rebase it on master now?

File: js2-imenu-extras.el [hunk 2]
```diff
@@ -221,6 +229,161 @@ Currently used for jQuery widgets, Dojo and Enyo declarations."
...
+  (js2-visit-ast
+   node
+   (lambda (child end-p)
+     (when (not end-p)
+       (js2-imenu-extras--mocha-check-unknown-node child qname)))))
+
+(defun js2-imenu-extras--mocha-check-unknown-node (node qname)
+  "If NODE is a mocha test block, populate `js2-imenu-recorder'.
+
+QNAME is a list of nodes representing the qualified name of
```
Comment (dgutov) [inline js2-imenu-extras.el:257]: The description of QNAME is repeated multiple times in other functions, the whole paragraph. Appreciate the good documentation, but it looks a little repetitive.

How about we keep it in `js2-imenu-extras--mocha-visit-node`, but in other functions say

    QNAME is the same as described in `js2-imenu-extras--mocha-visit-node'.

?

**Comment by dgutov** [comment]
A NEWS entry would be nice to have, too.

File: js2-imenu-extras.el (hunk 1 above)
Comment (DamienCassou) [reply js2-imenu-extras.el:236]: IMO, `(list)` better conveys the meaning of the parameter being a list than `nil`. I'm fine changing it to `nil` if you want me to.

**Comment by DamienCassou** [comment]
I took care of your feedback but there are issues with the tests in 25.1. Do you have any idea?

File: js2-imenu-extras.el (hunk 1 above)
Comment (dgutov) [reply js2-imenu-extras.el:236]: It's all right.

**Comment by dgutov** [comment]
> but there are issues with the tests in 25.1. Do you have any idea?

Looks like `when-let*` was not in Emacs 25. I think if you just change it to `when-let`, everything will work.

File: NEWS.md [hunk 3]
```diff
@@ -11,6 +11,10 @@
 * Support for trailing comma in arrow function parameters
   ([#480](https://github.com/mooz/js2-mode/issues/480)).
 * Support for `for await of` syntax ([#543](https://github.com/mooz/js2-mode/pull/543)).
+* Imenu support for mocha-like (includes Jasmine and Cypress) test
```
Comment (dgutov) [inline NEWS.md:14]: Please move the entry to the top: they all go in reverse-chronological order here.

Comment (DamienCassou) [reply NEWS.md]: fixed

**Comment by DamienCassou** [comment]
> Looks like when-let* was not in Emacs 25. I think if you just change it to when-let, everything will work.

Indeed. I also had to `require` "subr-x". I also added Emacs 26.3 to GitHub Actions because it also needs "subr-x".

File: js2-imenu-extras.el [hunk 4]
```diff
@@ -35,6 +35,10 @@
 (require 'cl-lib)
 (require 'js2-mode)
 
+(eval-when-compile
+  (when (<= emacs-major-version 26)
+    (require 'subr-x)))
```
Comment (dgutov) [inline js2-imenu-extras.el:40]: The version check wasn't really necessary, FWIW.

`when-let` is autoloaded in newer Emacs, but it doesn't hurt to require the package too.

**Comment by dgutov** [comment]
Thanks!

**Comment by DamienCassou** [comment]
Thank you for your reviews and the package!

**Pull request merged**